solidity/test/bridge/Bridge.test.ts

```diff
@@ -1423,6 +1448,704 @@ describe("Bridge", () => {
+                              )
+                              expect(
+                                bridgeBalance.sub(initialBridgeBalance)
+                              ).to.equal(requestedAmount)
+
+                              const redeemerBalance = await bank.balanceOf(
+                                redeemer.address
+                              )
+                              expect(
+                                redeemerBalance.sub(initialRedeemerBalance)
+                              ).to.equal(requestedAmount.mul(-1))
+                            })
+                          }
+                        )
+
+                        context(
+                          "when redeemer has not made a sufficient allowance in Bank",
+                          () => {
+                            it("should revert", async () => {
+                              // TODO: Implementation.
+                            })
+                          }
+                        )
+                      })
+
+                      context("when wallet has insufficient funds", () => {
+                        it("should revert", async () => {
+                          // TODO: Implementation.
+                        })
+                      })
+                    }
+                  )
+
+                  context(
+                    "when there is a pending request for the given redemption key",
+                    () => {
+                      it("should revert", async () => {
+                        // TODO: Implementation.
+                      })
+                    }
+                  )
+                })
+
+                context("when amount is below the dust threshold", () => {
+                  it("should revert", async () => {
+                    // TODO: Implementation.
+                  })
+                })
+              }
+            )
+
+            context(
+              "when redeemer output script points to the wallet public key hash",
+              () => {
+                it("should revert", async () => {
+                  // TODO: Implementation. Make sure there is not possibility
+                  //       to pass the 20-byte wallet PKH under P2PKH, P2WPKH
+                  //       and P2SH.
+                })
+              }
+            )
+          })
+
+          context("when redeemer output script is not standard type", () => {
+            it("should revert", async () => {
+              // TODO: Implementation.
+            })
+          })
+        })
+
+        context("when main UTXO data are invalid", () => {
+          it("should revert", async () => {
+            // TODO: Implementation.
+          })
+        })
+      })
+
+      context("when there is no main UTXO for the given wallet", () => {
+        it("should revert", async () => {
+          // TODO: Implementation.
+        })
+      })
+    })
+
+    context("when wallet state is other than active", () => {
```

Comment:
I think we should check other states separately and make sure no other state than `Active` is going to work. Unknown state should be covered by a separate test case for an unknown wallet.

Reply:
These will be separate sub-contexts within this one. I'll drop a TODO to make it explicit. https://github.com/keep-network/tbtc-v2/pull/128/commits/bd7b48fc10fcb38ee744a3df0676a2919130ab1c